### PR TITLE
Fix #2668 Increase precision of Query Panel Circle/BBox Details

### LINKED ISSUE #2668: Increase precision of Query Panel Circle/BBox Details
### Description
Query Panel Geometry details has a decimal precision fixed to 2 digits. It is not enough. User can not edit precise circle filter or bbox filters at high zoom levels. 

 - When coordinates are in 4326 (`useMapProjection === false` in `GeometryDetails`)
when you change the radius of a circle the center of the circle moves to the nearest point with coordinates multiple of 0.01. 
 - When you click on input arrows, the step of 1 degree is too big. It should be smaller (may depend from the current map resolution, but a simple solution can be to setting the step of 0.01). 
 - Because of live update of the geometry, if the user deletes a digit and insert another one, the geometry may jump here and there. We may introduce a debounce time (1 sec) for geometry update.

### In case of Bug (otherwise remove this paragraph)

*Browser Affected*
any

*Steps to reproduce*

- Open the map
- Add a vector layer to the map
- Search for "Colosseo" and select the first result
- Open feature grid, then filter
- Select circle filter and draw a circle on the map centered in the coliseum center
image

- Click on pencil to open filter's geometry details
- change radius value

*Expected Result*
- The circle only changes radius

*Current Result*
 - The circle moves it's center to the nearest point with coordinates multiple of 0.01 
Something like this:
image

### Other useful information (optional):
This is due to a wrong approximation + to fast live editing + wrong step. 
We should: 
 - [ ] increase precision in case of 4326.
 - [ ]  debounce user input to update with 1s of delay.
 - [ ]  Increase step for the input component at least to 0.01
 - [ ]  Specify the unit of measure of the circle radius (m)

### PR DESCRIPTION
**Pull request by kappu72**
## Description
If Query Panel Geometry details has proj 4326 the precision off numeric fields  is increased to 6th decimal, the step is related to the current scale, and the update is debounced by 1 sec.
## Issues
 - Fix #2668

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)

 - [x] Bugfix
 - [ ] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)
All numeric fields has tow decimal precision, and the step is always one

**What is the new behaviour?**
If proj is 4326, coordinates fields has 6 decimal precision, the step is related to the scale and the update is debounced by one sec. The radius has 2 decimal precision.
**Does this PR introduce a breaking change?** (check one with "x", remove the other)

 - [ ] Yes
 - [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: web/client/components/data/query/GeometryDetails.jsx [hunk 1]
```diff
@@ -168,28 +169,44 @@ class GeometryDetails extends React.Component {
...
-        } else if (this.props.type === "Circle") {
-            this.resetCircle();
-        }
-
+        this.resetGeom();
         this.props.onShowPanel(false);
     };
-
+    roundValue = (val, prec = 1000000) => Math.round(val * prec) / prec;
+    getStep = (zoom = 1) => {
```
Comment (mbarto) [inline web/client/components/data/query/GeometryDetails.jsx]: I think we can translate this into a formula, in any case we shouldn't use ifs

File: web/client/components/data/query/GeometryDetails.jsx [hunk 2]
```diff
@@ -168,28 +169,44 @@ class GeometryDetails extends React.Component {
...
+        }else if( zoom >= 6) {
+            return 1
+        }
+            return 10;
+    };
+    getStepCircle = (zoom, name) => {
+        const step = this.getStep(zoom);
+        return name === 'radius' && step * 100000 || step;
+    };
+    isWGS84 = () => (this.props.geometry || {}).projection === 'EPSG:4326' || !this.props.useMapProjection;
```
Comment (mbarto) [inline web/client/components/data/query/GeometryDetails.jsx]: I think we can use proj4 to know the uom of the projection, and generalize EPSG:4326 to uom === 'degrees'

File: web/client/components/data/query/GeometryDetails.jsx [hunk 3]
```diff
@@ -77,7 +79,7 @@ class GeometryDetails extends React.Component {
             projection: this.props.geometry.projection
         };
 
-        this.props.onChangeDrawingStatus("replace", undefined, "queryform", [geometry]);
+        this.props.onChangeDrawingStatus({geometry: [geometry]});
```
Comment (mbarto) [inline web/client/components/data/query/GeometryDetails.jsx]: is CHANGE_DRAWING_STATUS without a "command" (e.g. replace) supported?

File: web/client/components/data/query/enhancers/geometrydetails.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+*
+* This source code is licensed under the BSD-style license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+const {compose, createEventHandler, defaultProps} = require('recompose');
+
+const propsStreamFactory = require('../../../misc/enhancers/propsStreamFactory');
+
+
+const dataStreamFactory = $props => {
```
Comment (mbarto) [inline web/client/components/data/query/enhancers/geometrydetails.js]: Why do we need this?

Comment (kappu72) [reply web/client/components/data/query/enhancers/geometrydetails.js]: Done to debounce the change.

Comment (mbarto) [reply web/client/components/data/query/enhancers/geometrydetails.js]: it also changes the interface to call onChangeDrawingStatus, which is not immediately clear, look at the other question a bove

File: web/client/components/data/query/GeometryDetails.jsx (hunk 3 above)
Comment (kappu72) [reply web/client/components/data/query/GeometryDetails.jsx]: maybe was better to change the name, but isn't directly calling the onChangeDrawingStatus action. That action is debounced and then called by enhancer

File: web/client/components/data/query/enhancers/geometrydetails.js (hunk 4 above)
Comment (kappu72) [reply web/client/components/data/query/enhancers/geometrydetails.js]: se answer above

File: web/client/components/data/query/GeometryDetails.jsx (hunk 3 above)
Comment (mbarto) [reply web/client/components/data/query/GeometryDetails.jsx]: I mean, if you put a "transparent" debounce in the middle, it should be transparent, if I remove it the same code will work, without debouncing. We want to enhance, not to change.

File: web/client/components/data/query/enhancers/geometrydetails.js (hunk 4 above)
Comment (mbarto) [reply web/client/components/data/query/enhancers/geometrydetails.js]: I mean, if you put a "transparent" debounce in the middle, it should be transparent, if I remove it the same code will work, without debouncing. We want to enhance, not to change.

File: web/client/components/data/query/GeometryDetails.jsx (hunk 3 above)
Comment (mbarto) [reply web/client/components/data/query/GeometryDetails.jsx]: Actions have a standard interface, you are changing it here

File: web/client/components/data/query/GeometryDetails.jsx [hunk 5]
```diff
@@ -168,28 +169,42 @@ class GeometryDetails extends React.Component {
...
-        } else if (this.props.type === "Circle") {
-            this.resetCircle();
-        }
-
+        this.resetGeom();
         this.props.onShowPanel(false);
     };
-
+    getStep = (zoom = 1) => {
+        if ( zoom >= 21 ) {
```
Comment (mbarto) [inline web/client/components/data/query/GeometryDetails.jsx]: Can we translate this into a formula?

File: web/client/components/data/query/enhancers/geometrydetails.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+* Copyright 2018, GeoSolutions Sas.
+* All rights reserved.
+*
+* This source code is licensed under the BSD-style license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+const {compose, withHandlers} = require('recompose');
+const {debounce} = require("lodash");
+
+module.exports = compose(withHandlers((initProp) => {
```
Comment (mbarto) [inline web/client/components/data/query/enhancers/geometrydetails.js]: Can't we generalize this into a "debouncing" enhancer? I don't get the purpose of having single use enhancers in separate files

File: web/client/components/misc/enhancers/debounce.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+* All rights reserved.
+*
+* This source code is licensed under the BSD-style license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+const {withHandlers} = require('recompose');
+const {debounce} = require("lodash");
+const emptyFunc = () => {};
+/**
+ * This enhancer debounce an action or of a passed debounceTime
```
Comment (offtherailz) [inline web/client/components/misc/enhancers/debounce.js]: This enhancer de-bounce a method passed as prop of the given time.

**Pull request merged**